runtime/common/src/evm/fees.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// Much of this code is copied verbatim from Frontier
+// Copyright (c) 2020-2022 Parity Technologies (UK) Ltd.
+
+use core::marker::PhantomData;
+
+use cfg_primitives::constants::TREASURY_FEE_RATIO;
+use frame_support::traits::{
+	Currency, ExistenceRequirement, Imbalance, OnUnbalanced, SignedImbalance, WithdrawReasons,
+};
+use pallet_balances::Pallet as Balances;
+use pallet_evm::{AddressMapping, OnChargeEVMTransaction};
+use pallet_treasury::Pallet as Treasury;
+use sp_core::{H160, U256};
+use sp_runtime::{
+	traits::{UniqueSaturatedInto, Zero},
+	Perbill, Saturating,
+};
+
+use crate::fees::{NegativeImbalance, ToAuthor};
+
+type AccountIdOf<R> = <R as frame_system::Config>::AccountId;
+type Balance<R> = <Balances<R> as Currency<AccountIdOf<R>>>::Balance;
+type PositiveImbalance<R> = <Balances<R> as Currency<AccountIdOf<R>>>::PositiveImbalance;
+
+type Error<R> = pallet_evm::Error<R>;
+
+pub struct DealWithEVMFees<R>(PhantomData<R>);
+
+impl<R> OnChargeEVMTransaction<R> for DealWithEVMFees<R>
+where
+	R: pallet_balances::Config
+		+ pallet_treasury::Config
+		+ pallet_authorship::Config
+		+ pallet_evm::Config,
+	Treasury<R>: OnUnbalanced<NegativeImbalance<R>>,
+	Balances<R>: Currency<<R as frame_system::Config>::AccountId>,
+	U256: UniqueSaturatedInto<Balance<R>>,
+{
+	type LiquidityInfo = Option<NegativeImbalance<R>>;
+
+	fn withdraw_fee(who: &H160, fee: U256) -> Result<Self::LiquidityInfo, Error<R>> {
+		if fee.is_zero() {
+			return Ok(None);
+		}
+		let account_id = R::AddressMapping::into_account_id(*who);
+		Balances::<R>::withdraw(
+			&account_id,
+			fee.unique_saturated_into(),
+			WithdrawReasons::FEE,
+			ExistenceRequirement::AllowDeath,
+		)
+		.map(Some)
+		.map_err(|_| Error::<R>::BalanceLow)
+	}
+
+	fn correct_and_deposit_fee(
+		who: &H160,
+		corrected_fee: U256,
+		base_fee: U256,
+		already_withdrawn: Self::LiquidityInfo,
+	) -> Self::LiquidityInfo {
+		let Some(paid) = already_withdrawn else { return None };
```

Comment:
rustfmt also unfortunately does not handle `let else` properly yet, so we should probably only use it for small things like this. Otherwise we'll end up with a bunch of reformats when rustfmt finally gets updated